icu4j/main/classes/core/src/com/ibm/icu/impl/number/UsagePrefsHandler.java

```diff
@@ -0,0 +1,118 @@
+    private UnitsRouter fUnitsRouter;
+
+    public UsagePrefsHandler(ULocale locale, MeasureUnit inputUnit, String usage, MicroPropsGenerator parent) {
+        assert parent != null;
+
+        this.fParent = parent;
+        this.fUnitsRouter =
+                new UnitsRouter(MeasureUnitImpl.forIdentifier(inputUnit.getIdentifier()), locale.getCountry(), usage);
+    }
+
+    private static Precision parseSkeletonToPrecision(String precisionSkeleton) {
+        final String kSuffixPrefix = "precision-increment/";
+        if (!precisionSkeleton.startsWith(kSuffixPrefix)) {
+            throw new IllegalIcuArgumentException("precisionSkeleton is only precision-increment");
+        }
+
+        String skeleton = precisionSkeleton.substring(kSuffixPrefix.length());
+        String skeletons[] = skeleton.split("/");
+        BigDecimal num = new BigDecimal(skeletons[0]);
+        BigDecimal den =
+                skeletons.length == 2 ?
+                        new BigDecimal(skeletons[1]) :
+                        new BigDecimal("1");
+
+
+        return Precision.increment(num.divide(den, MathContext.DECIMAL128));
+    }
+
+    protected static void mixedMeasuresToMicros(List<Measure> measures, DecimalQuantity quantity, MicroProps micros) {
+        if (measures.size() > 1) {
+            // For debugging
+            assert (micros.outputUnit.getComplexity() == MeasureUnit.Complexity.MIXED);
+
+            // Check that we received measurements with the expected MeasureUnits:
+            List<MeasureUnit> singleUnits = micros.outputUnit.splitToSingleUnits();
+
+            assert measures.size() == singleUnits.size();
+
+            // Mixed units: except for the last value, we pass all values to the
+            // LongNameHandler via micros->mixedMeasures.
+            for (int i = 0, n = measures.size() - 1; i < n; i++) {
+                micros.mixedMeasures.add(measures.get(i));
+            }
+        }
+
+        // The last value (potentially the only value) gets passed on via quantity.
+        quantity.setToBigDecimal((BigDecimal) measures.get(measures.size()- 1).getNumber());
+    }
+
+    /**
+     * Returns the list of possible output units, i.e. the full set of
+     * preferences, for the localized, usage-specific unit preferences.
+     * <p>
+     * The returned pointer should be valid for the lifetime of the
+     * UsagePrefsHandler instance.
+     */
+    public List<MeasureUnit> getOutputUnits() {
+        return fUnitsRouter.getOutputUnits();
+    }
+
+    /**
+     * Obtains the appropriate output value, MeasureUnit and
+     * rounding/precision behaviour from the UnitsRouter.
+     * <p>
+     * The output unit is passed on to the LongNameHandler via
+     * micros.outputUnit.
+     */
+    @Override
+    public MicroProps processQuantity(DecimalQuantity quantity) {
+        MicroProps micros = this.fParent.processQuantity(quantity);
+
+        quantity.roundToInfinity(); // Enables toDouble
```

Comment:
Is this needed for toBigDecimal() though? (Unit tests pass even when I comment this out.) I don't fully grasp DecimalQuantity's rounding yet.

Reply:
Leave it in.  We will solve the riddle later.